src/logic/__tests__/validateProject.test.js

```diff
@@ -0,0 +1,80 @@
+import {
+  projectRestoredFromLastSession as projectRestoredFromLastSessionAction,
+  snapshotImported as snapshotImportedAction,
+} from '../../actions/clients';
+import {validatedSource} from '../../actions/errors';
+import {
+  changeCurrentProject as changeCurrentProjectAction,
+  gistImported as gistImportedAction,
+  toggleLibrary as toggleLibraryAction,
+  updateProjectSource as updateProjectSourceAction,
+} from '../../actions/projects';
+import validateProject from '../validateProject';
+
+import {applyActions, makeTestLogic} from './helpers';
+
+import {firebaseProjectFactory} from '@factories/data/firebase';
+import {consoleErrorFactory} from '@factories/validations/errors';
+
+jest.mock('../../analyzers');
+
+const mockCssValidationErrors = [
+  consoleErrorFactory.build({
+    text: 'You have a starting { but no ending } to go with it.',
+  }),
+];
+
+const mockHtmlValidationErrors = [
+  consoleErrorFactory.build({
+    text: 'Closing tag missing',
+  }),
+];
+
+jest.mock('../../validations', () => ({
+  css: jest.fn(() => mockCssValidationErrors),
+  html: jest.fn(() => mockHtmlValidationErrors),
+  javascript: jest.fn(() => []),
+}));
+
+test('should validate current project', async () => {
+  const testLogic = makeTestLogic(validateProject);
+
+  [
+    changeCurrentProjectAction,
+    gistImportedAction,
+    snapshotImportedAction,
+    projectRestoredFromLastSessionAction,
+    toggleLibraryAction,
+  ].forEach(async action => {
```

Comment:
Having an un-awaited `async` function as the iterator is going to prevent the assertions from being executed before the test completes—I would recommend iterating over the actions _outside_ the `test` declaration, e.g.

```js
for (const action of [
  changeCurrentProjectAction,
  gistImportedAction,
  // ...
]) {
  test(`validates current project on ${action}`, async () => {
    // …
  }
}
```

Reply:
makes sense! will fix!